### ISSUE TITLE
I believe I've been able to improve some of the intellisense for model queries, please implement this (but maybe expand it first)

### ISSUE THREAD

**Issue by Evertt** [barryvdh/laravel-ide-helper#1536]
So I'm using VSCode with every extension I could install that gives my IDE the best possible understanding of how most of Laravel's magic methods work and what types they return. However, it's still not enough and luckily this package makes the autocompletion in my IDE better still.

However, I still found a scenario where my IDE doesn't understand what a certain method-call is referring to. Let me show you an example.

I have this model:

```php
namespace App\Models;

use Illuminate\Database\Eloquent\Model;

/**
 * @mixin IdeHelperTrainingDate
 */
class TrainingDate extends Model
{
    public function scopeSearch($query, $search) {}
}
```

And then this in my `_ide_helper_models.php` file:

```php
namespace App\Models{
/**
 * App\Models\TrainingDate
 * 
 * @method static \Illuminate\Database\Eloquent\Builder|TrainingDate query()
 * @method static \Illuminate\Database\Eloquent\Builder|TrainingDate search($search)
 * @mixin \Eloquent
 */
    class IdeHelperTrainingDate {}
}
```

Strictly speaking, I don't even need the line for `search` there, because one of my extensions is already smart enough to be able to link `search()` to `scopeSearch()`, but it's fine. So when I do this, things work completely fine:

```php
$training_dates = TrainingDate::query()
    ->search("foo")
    ->get();
```

I can hover the `search()` call, and it will give me the desired type information and I can **cmd + click** on it and it will refer me directly to `scopeSearch()`. However, that breaks as soon as I do this:

```php
$training_dates = TrainingDate::query()
    ->with("training")
    ->search("foo")
    ->get();
```

Because `with()` simply returns `Builder` and now neither my extensions nor your `_ide_helper_models.php` file is able to make my IDE understand that `search()` is still available. I thought maybe running `php artisan ide-helper:eloquent` and / or `php artisan ide-helper:generate` would fix that issue, but it didn't. However, I've been able to fix it myself! 😃 By adding the following lines at the top of `_ide_helper_models.php`:

```php
namespace Illuminate\Database\Eloquent {
	/**
	 * @template TModel of \Eloquent
	 * @mixin TModel
	 */
	class Builder {}
}
```

And then changing docblock of IdeHelperTrainingDate to the following:

```php
namespace App\Models{
/**
 * App\Models\TrainingDate
 * 
 * @method static \Illuminate\Database\Eloquent\Builder<TrainingDate> query()
 * @method static \Illuminate\Database\Eloquent\Builder<TrainingDate> search($search)
 * @mixin \Eloquent
 */
    class IdeHelperTrainingDate {}
}
```

Now I can easily mix and match methods coming from `Builder` and scope methods coming from `TrainingDate` and my IDE will never forget that it can keep using both.

Now of course, this was the quickest fix I was able to do, because I'm on a deadline and I'm sure there are lots of caveats here that I haven't thought of yet. And maybe the intellisense of IDEs could be expanded even more through methods similar to these.

So basically my request is, would you be willing to check out whether this approach is viable. As in, whether this doesn't create more problems than it solves. And if it does seem like a viable approach to extend IDE's intellisense then would you be willing to experiment what else could be improved with methods like these?

**Comment by mreduar** [comment]
Hello @Evertt Have you found a solution?

Please can you tell me more about the extension you use to detect the scopes without the with?
> Strictly speaking, I don't even need the line for search there, because one of my extensions is already smart enough to be able to link search() to scopeSearch(), but it's fine. So when I do this, things work completely fine:

**Comment by Evertt** [comment]
@mreduar Well I'm starting to think that I should just completely and only rely on phpstan with larastan.

> Please can you tell me more about the extension you use to detect the scopes without the with?

It seems I was mistaken. I just turned off the extension that I thought was making that work (which was [this one](https://marketplace.visualstudio.com/items?itemName=amiralizadeh9480.laravel-extra-intellisense), but it turns out that extension was not making the making possible, cause it still worked when I turned it off. Then I thought maybe phpstan / larastan was doing it, but I disabled those too and then it still worked. Turns out that somehow `_ide_helper_models.php` was making it work, because it stopped working when I deleted that file, even though I don't completely understand how. Because I can't see how that file, which defines the scoped method simply like `@method static \Illuminate\Database\Eloquent\Builder|TrainingDate search($search)`, can make VSCode understand that that refers to the `scopeSearch` method on the model.

I think I'm making things very confusing for myself by having so many extensions and packages installed to improve VSCode's intellisense and now I find it hard determining what extension or package is doing what exactly.
